File: services/rust/actix/src/routes/handlers/get.rs
use crate::routes::{errors::AppError, extractors::AppPath};
use actix_web::{HttpResponse, ResponseError, web::Data};
use app::state::AppState;
use std::sync::Mutex;
use tracing::{error, info, instrument, warn};

#[instrument(skip_all)]
pub async fn get_handler(state: Data<Mutex<AppState>>, AppPath(id): AppPath) -> HttpResponse {
    if let Ok(state) = state.lock() {
        if let Some(task) = state.tasks.get(&id).cloned() {
            drop(state);
            info!(
                %id,
                secret = task.secret.len(),
                operation = task.operation.to_string(),
                method = "GET",
                "Retrieved task"
            );

            return HttpResponse::Ok().json(task);
        } else {
            drop(state);
            warn!(%id, method = "GET", "Task not found");
            return AppError::TaskNotFound.error_response();
        }
    }

    error!(%id, method = "GET", "Poisoned lock");

    AppError::Internal.error_response()
}

File: services/go/gin/routes/handlers/get.go
package handlers

import (
	"app"
	"gin/routes"
	"log/slog"
	"net/http"
	"strings"

	"github.com/gin-gonic/gin"
)

func GetHandler(state *app.AppState) gin.HandlerFunc {
	return func(c *gin.Context) {
		id, err := ParseID(c)
		if err != nil {
			return
		}

		state.Mu.RLock()
		task, ok := state.Tasks[id]
		state.Mu.RUnlock()

		if !ok {
			slog.Warn("Task not found", "method", c.Request.Method, "path", c.Request.URL.Path, "id", id.String())
			routes.AppErrors.TaskNotFound.Error(c)
			return
		}

		slog.Info(
			"Retrieved task",
			"id", id.String(),
			"secret", len(task.Secret),
			"operation", strings.ToLower(string(task.Operation)),
			"method", c.Request.Method,
		)

		c.JSON(http.StatusOK, task)
	}
}
